internal/topology/upgrade/clusterctl_upgrade_test.go

```diff
@@ -0,0 +1,1174 @@
+			Name:     refObj.GetName(),
+		}] = refObj.GetGeneration()
+		if additionalChecks != nil {
+			if err := additionalChecks(refObj); err != nil {
+				return nil, errors.Wrap(err, "failed additional checks on machineDeployment.spec.template.spec.infrastructureRef")
+			}
+		}
+
+		if md.Spec.Template.Spec.Bootstrap.ConfigRef == nil {
+			return nil, errors.New("cluster machineDeployment.spec.template.spec.bootstrap.configRef is not yet set")
+		}
+		refObj, err = getReferencedObject(ctx, env.GetClient(), md.Spec.Template.Spec.Bootstrap.ConfigRef, version, actualCluster.Namespace)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get referenced machineDeployment.spec.template.spec.bootstrap.configRef")
+		}
+		addToAllObj(cluster.Name, "machineDeployment "+md.Name+" .spec.template.spec.bootstrap.configRef", refObj)
+		refs[clusterv1.ContractVersionedObjectReference{
+			APIGroup: refObj.GroupVersionKind().Group,
+			Kind:     refObj.GetKind(),
+			Name:     refObj.GetName(),
+		}] = refObj.GetGeneration()
+		if additionalChecks != nil {
+			if err := additionalChecks(refObj); err != nil {
+				return nil, errors.Wrap(err, "failed additional checks on machineDeployment.spec.template.spec.bootstrap.configRef")
+			}
+		}
+	}
+	return refs, nil
+}
+
+func getReferencedObject(ctx context.Context, c client.Client, ref *clusterv1.ContractVersionedObjectReference, version, namespace string) (*unstructured.Unstructured, error) {
+	refObj := &unstructured.Unstructured{}
+	refObj.SetGroupVersionKind(ref.GroupKind().WithVersion(version))
+	if err := c.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: namespace}, refObj); err != nil {
+		return nil, err
+	}
+	return refObj, nil
+}
+
+func checkOmitableField(allowOmitable bool) func(obj *unstructured.Unstructured) error {
+	return func(obj *unstructured.Unstructured) error {
+		switch obj.GetKind() {
+		case "TestResource":
+			_, exists, err := unstructured.NestedFieldNoCopy(obj.Object, "spec", "omitable")
+			if err != nil {
+				return err
+			}
+			if exists && !allowOmitable {
+				return errors.Errorf("expected to not contain omitable field")
+			}
+		case "TestResourceTemplate":
+			_, exists, err := unstructured.NestedFieldNoCopy(obj.Object, "spec", "template", "spec", "omitable")
+			if err != nil {
+				return err
+			}
+			if exists && !allowOmitable {
+				return errors.Errorf("expected to not contain omitable field")
+			}
+		}
+		return nil
+	}
+}
+
+// assertRollout assert a rollout happened by checking referenced template are changed or referencedObjects increased generation.
+func assertRollout(g *WithT, cluster *clusterv1.Cluster, refsBefore, refsAfter map[clusterv1.ContractVersionedObjectReference]int64) {
+	actualCluster := &clusterv1.Cluster{}
+	err := env.Get(ctx, client.ObjectKeyFromObject(cluster), actualCluster)
+	g.Expect(err).To(Succeed())
+
+	clusterRef := clusterv1.ContractVersionedObjectReference{
+		APIGroup: actualCluster.GroupVersionKind().Group,
```

Comment:
Same comment as above about gvk in typed objects, please also check other cases below

(I"m mentioning this because every CR update could break this, not sure IIRC but the next one might actually break it)